src/compiler/checker.ts

```diff
@@ -8266,18 +8301,38 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return { introducesError, node: attachSymbolToLeftmostIdentifier(node) as T };
             }
             sym = resolveEntityName(leftmost, meaning, /*ignoreErrors*/ true, /*dontResolveAlias*/ true);
+            if (sym === unknownSymbol) {
+                // When we create a fake scope some parameters may actually not be usable
+                // either because they are the expanded rest parameter,
+                // or because they are the newly added parameters from the tuple, which might have different meanings in teh original context
+                // So we add unknownSymbol for these, and we fall back to inference when we encounter them
+                introducesError = true;
+                return { introducesError, node, sym };
+            }
+            if (
+                context.enclosingDeclaration &&
+                (getNodeLinks(context.enclosingDeclaration).fakeScopeForSignatureDeclaration || !findAncestor(node, n => n === context.enclosingDeclaration)) &&
+                !(sym && sym.flags & SymbolFlags.TypeParameter)
+            ) {
+                const symAtLocation = resolveEntityName(leftmost, meaning, /*ignoreErrors*/ true, /*dontResolveAlias*/ true, context.enclosingDeclaration);
+                if ((symAtLocation !== sym && (!symAtLocation || !isTransientSymbol(symAtLocation) || symAtLocation.links.target !== sym)) || symAtLocation === unknownSymbol) {

```

Comment:
The later half of this condition needs some explanation - both here and in a comment, because I'm not sure offhand why we're looking at the cached target (and why that's safe) and the transient-ness of the symbol to determine if the resolution is an error. I'd think any symbol that isn't "the same" (eg, `getSymbolIfSameReference`) would be an error.

Also: If these conditions are met, isn't this the only lookup we need to do? Doesn't matter if it resolves at the original location if it resolves in the new one, no? Not sure how that could happen, but we could cut down on name resolutions we don't need to do.

Reply:
The symbol resolved in the current enclosing scope might be different from the declaration scope since we only really look at the name. In https://github.com/microsoft/TypeScript/issues/58407 we see how this could happen even today when we are a lot more conservative about what declarations we transplant at a new location.


The `isTransientSymbol` is indeed superfluous. I removed it and used `getSymbolIfSameReference` instead to test for symbol sameness. That revealed some instances where the symbol was indeed the same that I was missing before. Thank you. 

I was wondering, `getSymbolIfSameReference` does not check `s1 === s2` as a fast path. would that make sense?  I am assuming most of the symbols will be resolved so probably there is no actual lookup happening, just looking at cached values.

Reply:
I added the explanation